### ISSUE TITLE
Errored requests wipe the data: change that behavior?

### ISSUE THREAD

**Issue by jamesplease** [jamesplease/react-request#154]
Right now, the following occurs:

1. Fetch some data in a component
2. Make a request
3. Abort the request by immediately changing props or calling `doFetch` again
4. Observe that the data that is passed to the render prop is set to `null` when the cancel error is passed down

I think it is worth considering if the previous data should stick around when requests are aborted. This would be a breaking change.

The reason I am in favor of this is that if you call `doFetch()` 10 times in a row, there shouldn't be any flash of white when the data disappears. It should just continue to show what used to be there.

This likely applies to all errors.

The current behavior is verified as part of the `doFetch()` dedupe tests.

**Comment by janbaykara** [comment]
There are two issues:

#### 1. Data during (re)fetch on a key
Provide a `isSeamless` prop that determines whether to wipe data between requests. Should prolly default to `true` to satisfy your x10 in a row example.

#### 2. Data after error
Provide a `persistOnError` prop that determines whether to wipe data on error. I'd argue this one should default to `false` since it is the most honest outcome of a fetch request.

**Comment by jamesplease** [comment]
Thanks for the suggestions, @janbaykara ! I agree with you that those are two separate concerns.

Right now, I am having trouble thinking up a use case for `isSeamless` being `false`. If you have one, or can think of one, then we would definitely consider adding it ✌️

The `persistOnError` idea has some benefits, too. It would be backwards compatible if its `false`, and it gives consumers more control.

However, adhering to the behavior of `fetch` is not a primary intention for this library, because the “raw behavior” of fetch was not designed to make developer’s lives as easy as possible. They are just trying to map an HTTP request lifecycle to a Promise, and I think they are doing a good job at it, but using `fetch` directly is not always as _expressive_ as it could be.

One example is that the `failed` Boolean that you receive in the render prop is a combination of checking for a rejected Fetch promise and a resolved Fetch promise, because that’s how most developers actually think about unsuccessful requests. In fact, adding that boolean was a request from some developers at work who were sick of writing the longer code to check for `failed` all over the place 😅 

tl;dr, the design goals for this lib are:

- don’t lose any features of `fetch`, and add features when it makes sense (such as response caching and request de duplication)
- provide a different API when it lines up more with how developers will most likely want to make HTTP requests

Anyway, I’m writing all of this because I’ve chatted with a few folks at Netflix, and the current thought was that the data would just stick around after a failed request. What’s neat is that it gives you all of the power of both values of `persistOnError` without introducing the prop:

1. if you want `persistOnError: true`, then you can use the `data` prop provided to you to to continue rendering your UI (this is what most of our apps will do at Netflix, and I think that most other apps would also continue doing this).

2. if you want `persistOnError: false`, then you can check for the aforementioned `failed` Boolean, and choose not to use `data`.

Do you think this behavior would work for your use case @janbaykara , or do you still think that the prop make more sense?

Thanks for taking the time to share your thoughts!

**Comment by janbaykara** [comment]
I see your point, `persistOnError` is redundant.

With `isSeamless`, I once built a chart component that switched between different data views (say, `{ x: Month, y: ValueA }` and `{ x: Day, y: ValueB }`). Toggling between the two would result in a re-fetch for the new data, but since the *state* had already changed, the new subtree was expecting data for that new state, but it was instead receiving the data for the seamlessly provided previous fetch. In this case, I added `isSeamless={false}` to wipe the subtree slate clear.

But once again, I could have just listened for `fetching` regardless of the current `data` values.

**Comment by jamesplease** [comment]
@janbaykara , that use case for `isSeamless` makes sense to me. I am thinking that the same thing could be accomplished here by looking at `fetching` and possibly comparing the `url` as well (if you needed to differentiate switching between different views or refreshing the current data).

If you need to do something like that in the future, and this library doesn't support it, definitely open a new issue about that, and I will make sure that it gets fixed.

Since this original issue was about errors, I'm going to close this out as being resolved by #168 . It will be a part of `react-request@3.0.0`.
